### ISSUE TITLE
Guile Scheme support 

### ISSUE THREAD

**Issue by UrNightmaree** [6cdh/tree-sitter-scheme#6]
Will there be a support for Guile Scheme?

**Comment by 6cdh** [comment]
I don't have plan to support any specific implementation unless it conforms to the RnRS standard.
However, I'm considering make this project generate different parsers with a given flag to support common implementations.
What features do you need from guile scheme?

**Comment by archseer** [comment]
Maybe the parser could at least be more permissive (or skip certain tokens)?

https://git.savannah.gnu.org/cgit/guile.git/tree/module/language/tree-il/spec.scm

This file fails to highlight for me completely because of tokens like `#:use-module`. It doesn't seem like error recovery is working correctly, I'd expect it to skip the token and parse the rest of the file

**Comment by 6cdh** [comment]
it's weird. `#:xxx` is designed to be recognized to be a symbol. So people can use queries to make it be a keyword when they need.

**Comment by archseer** [comment]
Parsing somehow breaks for me so that each following define becomes just a single symbol:

![bug](https://github.com/6cdh/tree-sitter-scheme/assets/1372918/8c05450b-9381-40cc-9f30-8cd835c1833a)

**Comment by archseer** [comment]
Could be a bug in the latest tree-sitter master though, let me revert back to last release

**Comment by archseer** [comment]
Yeah sorry about that, it's a tree-sitter bug

**Comment by 6cdh** [comment]
I also found that feature that `#:xxx` is parsed as a symbol no longer works after a refactor.
It's fine because only RnRS standard should be implemented currently.

**Comment by archseer** [comment]
I think it's good if grammars are more permissive rather than being strict to the spec, that's the stance being taken on official grammars (https://github.com/tree-sitter/tree-sitter/issues/1631#issuecomment-1026833336). Since tree-sitter is used to implement editor tooling rather than parsing/compiling it's OK if it's a bit more forgiving.

**Comment by 6cdh** [comment]
Yeah. My naive idea was to support all implementation when this project started.
However, then I found it's impossible due to conflicting grammars of different scheme implementations.

I set a boundary between what should be implemented and what not.
Currently the boundary is RnRS standard. Anything beyond is not be considered.

The boundary is necessary to ensure the parser has a relatively strict grammar and deterministic.

Tree sitter is a convenient framework. But it's also very easy to have indeterminate behavior because of its design.
For example, some tree sitter parsers always have unexpected bugs. Even the authors don't know the behavior of their parsers.

While tree-sitter-scheme is quite stable, fast, and has fewer bugs. It only uses a limited set of tree-sitter features (for example, no `extras`).
And I have a clear understanding of its capabilities.

To expand the boundary, we need to carefully consider the popular implementations and their popular features.
Then strike a balance between them, and implement them.

Another possible design is to generate different parsers for different implementations, and the main parser is only standard.
But it has extra costs for the users.
